dotnet/src/MicroFocus.CafApi.QueryBuilder.Matcher/MatcherQueryVisitor.cs

```diff
@@ -145,83 +241,228 @@ public void VisitBetween(IMatcherFieldSpec<Document> fieldSpec, string startValu
+            if (values == null)
+            {
+                _isMatch = null;
+            }
+            else
+            {
+                string valueToCompare = fieldSpec.IsCaseInsensitive ? value.ToUpperInvariant() : value;
 
-            _isMatch = GetStringValues(fieldSpec).Any(v => v.CompareTo(valueToCompare) < 1);
+                _isMatch = values.Any(v => v.CompareTo(valueToCompare) < 1);
+            }
         }
 
         public void VisitGreaterThan(IMatcherFieldSpec<Document> fieldSpec, long value)
         {
-            _isMatch = GetLongStream(fieldSpec).Any(v => v > value);
+            var values = GetLongStream(fieldSpec);
+            if (values == null)
+            {
+                _isMatch = null;
+            }
+            else
+            {
+                _isMatch = values.Any(v => v > value);
+            }
         }
 
         public void VisitGreaterThan(IMatcherFieldSpec<Document> fieldSpec, string value)
         {
-            string valueToCompare = fieldSpec.IsCaseInsensitive ? value.ToUpperInvariant() : value;
+            var values = GetStringValues(fieldSpec);
+            if (values == null)
+            {
+                _isMatch = null;
+            }
+            else
+            {
+                string valueToCompare = fieldSpec.IsCaseInsensitive ? value.ToUpperInvariant() : value;
 
-            _isMatch = GetStringValues(fieldSpec).Any(v => v.CompareTo(valueToCompare) > 0);
+                _isMatch = values.Any(v => v.CompareTo(valueToCompare) > 0);
+            }
         }
 
         public void VisitGreaterThanOrEquals(IMatcherFieldSpec<Document> fieldSpec, long value)
         {
-            _isMatch = GetLongStream(fieldSpec).Any(v => v >= value);
+            var values = GetLongStream(fieldSpec);
+            if (values == null)
+            {
+                _isMatch = null;
+            }
+            else
+            {
+                _isMatch = values.Any(v => v >= value);
+            }
         }
 
         public void VisitGreaterThanOrEquals(IMatcherFieldSpec<Document> fieldSpec, string value)
         {
-            string valueToCompare = fieldSpec.IsCaseInsensitive ? value.ToUpperInvariant() : value;
+            var values = GetStringValues(fieldSpec);
+            if (values == null)
+            {
+                _isMatch = null;
+            }
+            else
+            {
+                string valueToCompare = fieldSpec.IsCaseInsensitive ? value.ToUpperInvariant() : value;
 
-            _isMatch = GetStringValues(fieldSpec).Any(v => v.CompareTo(valueToCompare) >= 0);
+                _isMatch = values.Any(v => v.CompareTo(valueToCompare) >= 0);
+            }
         }
 
         public void VisitExists(IMatcherFieldSpec<Document> fieldSpec)
         {
-            _isMatch = GetFieldValues(fieldSpec).Any();
+            var values = GetFieldValues(fieldSpec);
+            if (values == null)
+            {
+                _isMatch = false;
```

Comment:
Debatable I know but I think this should be `null`.

```suggestion
                _isMatch = null;
```